File: app/src/main/java/me/saket/dank/data/ErrorResolver.java
package me.saket.dank.data;

import android.support.annotation.Nullable;

import com.bumptech.glide.load.engine.GlideException;

import java.io.InterruptedIOException;
import java.net.SocketException;
import java.net.SocketTimeoutException;
import java.net.UnknownHostException;
import java.util.concurrent.CancellationException;
import java.util.concurrent.ExecutionException;
import javax.inject.Inject;

import io.reactivex.exceptions.CompositeException;
import io.reactivex.exceptions.UndeliverableException;
import me.saket.dank.R;
import me.saket.dank.data.exceptions.ImgurApiRequestRateLimitReachedException;
import me.saket.dank.data.exceptions.ImgurApiUploadRateLimitReachedException;
import me.saket.dank.utils.Optional;
import retrofit2.HttpException;

/**
 * Resolves commonly encountered errors and and constructs a user understandable
 * title and message that can be shown by the View.
 */
public class ErrorResolver {

  @Inject
  public ErrorResolver() {
  }

  public ResolvedError resolve(@Nullable Throwable error) {
    error = findActualCause(error);

    if (isNetworkTimeoutError(error)) {
      if (error instanceof UnknownHostException
          && error.getMessage() != null
          && (error.getMessage().contains("reddit.com") || error.getMessage().contains("redditmedia.com")))
      {
        return ResolvedError.create(
            ResolvedError.Type.NETWORK_ERROR,
            R.string.common_network_error_emoji,
            R.string.common_network_error_with_reddit_message);

      } else {
        return ResolvedError.create(
            ResolvedError.Type.NETWORK_ERROR,
            R.string.common_network_error_emoji,
            R.string.common_network_error_with_other_websites_message);
      }

    } else if (error instanceof HttpException && ((HttpException) error).code() == 503) {
      return ResolvedError.create(
          ResolvedError.Type.REDDIT_IS_DOWN,
          R.string.common_reddit_is_down_error_emoji,
          R.string.common_reddit_is_down_error_message);

    } else if (error instanceof ImgurApiRequestRateLimitReachedException) {
      return ResolvedError.create(
          ResolvedError.Type.IMGUR_RATE_LIMIT_REACHED,
          R.string.common_imgur_rate_limit_error_emoji,
          R.string.common_imgur_request_rate_limit_error_message);

    } else if (error instanceof ImgurApiUploadRateLimitReachedException) {
      return ResolvedError.create(
          ResolvedError.Type.IMGUR_RATE_LIMIT_REACHED,
          R.string.common_imgur_rate_limit_error_emoji,
          R.string.common_imgur_upload_rate_limit_error_message);

    } else if (error instanceof CancellationException || error instanceof InterruptedIOException) {
      return ResolvedError.create(
          ResolvedError.Type.CANCELATION,
          R.string.common_error_cancelation_emoji,
          R.string.common_error_cancelation_message);

    } else if (error instanceof IllegalStateException
        && Optional.ofNullable(error.getMessage()).map(m -> m.contains("no active user context")).orElse(false))
    {
      return ResolvedError.create(
          ResolvedError.Type.UNKNOWN,
          R.string.common_error_known_jraw_emoji,
          R.string.common_error_known_jraw_message);

    } else {
      return ResolvedError.create(
          ResolvedError.Type.UNKNOWN,
          R.string.common_unknown_error_emoji,
          R.string.common_unknown_error_message);
    }
  }

  public Throwable findActualCause(@Nullable Throwable error) {
    if (error instanceof ExecutionException) {
      error = findActualCause(error.getCause());
    }
    if (error instanceof GlideException && !((GlideException) error).getRootCauses().isEmpty()) {
      error = ((GlideException) error).getRootCauses().get(0);
    }
    if (error instanceof CompositeException) {
      error = findActualCause(error.getCause());
    }
    if (error instanceof UndeliverableException) {
      error = findActualCause(error.getCause());
    }
    if (error instanceof RuntimeException && error.getCause() != null && isNetworkTimeoutError(error.getCause())) {
      // Stupid JRAW wraps all HTTP exceptions with RuntimeException.
      error = findActualCause(error.getCause());
    }
    if (error instanceof IllegalStateException && error.getMessage() != null && error.getMessage().contains("Reached retry limit")) {
      error = findActualCause(error.getCause());
    }
    return error;
  }

  private boolean isNetworkTimeoutError(@Nullable Throwable error) {
    return error instanceof SocketException || error instanceof SocketTimeoutException || error instanceof UnknownHostException;
  }
}

File: app/src/main/java/me/saket/dank/data/ErrorResolver.kt
package me.saket.dank.data

import com.bumptech.glide.load.engine.GlideException
import io.reactivex.exceptions.CompositeException
import io.reactivex.exceptions.UndeliverableException
import me.saket.dank.R
import me.saket.dank.data.exceptions.ImgurApiRequestRateLimitReachedException
import me.saket.dank.data.exceptions.ImgurApiUploadRateLimitReachedException
import okhttp3.internal.http2.ConnectionShutdownException
import okhttp3.internal.http2.StreamResetException
import retrofit2.HttpException
import java.io.FileNotFoundException
import java.io.InterruptedIOException
import java.net.SocketException
import java.net.SocketTimeoutException
import java.net.UnknownHostException
import java.util.concurrent.CancellationException
import java.util.concurrent.ExecutionException
import javax.inject.Inject

/**
 * Resolves commonly encountered errors and and constructs a user understandable
 * title and message that can be shown by the View.
 */
class ErrorResolver @Inject
constructor() {

  fun resolve(error: Throwable?): ResolvedError {
    if (error == null) {
      return ResolvedError.create(
          ResolvedError.Type.UNKNOWN,
          R.string.common_unknown_error_emoji,
          R.string.common_unknown_error_message)
    }

    val actualError = findActualCause(error)

    return if (isNetworkTimeoutError(actualError)) {
      if (actualError is UnknownHostException
          && actualError.message != null
          && (actualError.message!!.contains("reddit.com") || actualError.message!!.contains("redditmedia.com"))) {
        ResolvedError.create(
            ResolvedError.Type.NETWORK_ERROR,
            R.string.common_network_error_emoji,
            R.string.common_network_error_with_reddit_message)

      } else {
        ResolvedError.create(
            ResolvedError.Type.NETWORK_ERROR,
            R.string.common_network_error_emoji,
            R.string.common_network_error_with_other_websites_message)
      }

    } else if (actualError is HttpException && actualError.code() == 503) {
      ResolvedError.create(
          ResolvedError.Type.REDDIT_IS_DOWN,
          R.string.common_reddit_is_down_error_emoji,
          R.string.common_reddit_is_down_error_message)

    } else if (actualError is ImgurApiRequestRateLimitReachedException) {
      ResolvedError.create(
          ResolvedError.Type.IMGUR_RATE_LIMIT_REACHED,
          R.string.common_imgur_rate_limit_error_emoji,
          R.string.common_imgur_request_rate_limit_error_message)

    } else if (actualError is ImgurApiUploadRateLimitReachedException) {
      ResolvedError.create(
          ResolvedError.Type.IMGUR_RATE_LIMIT_REACHED,
          R.string.common_imgur_rate_limit_error_emoji,
          R.string.common_imgur_upload_rate_limit_error_message)

    } else if (actualError is CancellationException || actualError is InterruptedIOException || actualError is InterruptedException) {
      ResolvedError.create(
          ResolvedError.Type.CANCELATION,
          R.string.common_error_cancelation_emoji,
          R.string.common_error_cancelation_message)

    } else if (actualError is IllegalStateException && actualError.message?.contains("no active user context") == true) {
      ResolvedError.create(
          ResolvedError.Type.KNOWN_BUT_IGNORED,
          R.string.common_error_known_jraw_emoji,
          R.string.common_error_known_jraw_message)

    } else if (actualError is FileNotFoundException && actualError.message?.contains("No content provider") == true) {
      ResolvedError.create(
          ResolvedError.Type.KNOWN_BUT_IGNORED,
          R.string.common_error_cancelation_emoji,
          R.string.common_error_cancelation_message)

    } else {
      ResolvedError.create(
          ResolvedError.Type.UNKNOWN,
          R.string.common_unknown_error_emoji,
          R.string.common_unknown_error_message)
    }
  }

  fun findActualCause(error: Throwable): Throwable {
    var actualError = error

    if (actualError is ExecutionException) {
      // AFAIK, thrown by Glide in situations like socket-timeout.
      actualError = findActualCause(actualError.cause!!)
    }
    if (actualError is GlideException && actualError.rootCauses.isNotEmpty()) {
      actualError = actualError.rootCauses.last()
    }
    if (actualError is CompositeException) {
      actualError = findActualCause(actualError.cause)
    }
    if (actualError is UndeliverableException) {
      actualError = findActualCause(actualError.cause!!)
    }
    if (actualError is RuntimeException && actualError.cause != null && isNetworkTimeoutError(actualError.cause)) {
      // Stupid JRAW wraps all HTTP exceptions with RuntimeException.
      // Update: this may no longer be true with JRAW v1.0 (Dank v.0.6.1).
      actualError = findActualCause(actualError.cause!!)
    }
    if (actualError is IllegalStateException && actualError.message?.contains("Reached retry limit") == true) {
      actualError = findActualCause(actualError.cause!!)
    }
    return actualError
  }

  private fun isNetworkTimeoutError(error: Throwable?): Boolean {
    return (error is SocketException
        || error is SocketTimeoutException
        || error is UnknownHostException
        || error is StreamResetException
        || error is ConnectionShutdownException)
  }
}
